plugins/thunderbolt-power/fu-plugin-thunderbolt-power.c

```diff
@@ -0,0 +1,274 @@
+		g_set_error (error, G_IO_ERROR,
+			     g_io_error_from_errno (errno),
+			     "could not write to force_power': %s",
+			     g_strerror (errno));
+		g_close (fd, NULL);
+		return FALSE;
+	}
+	return g_close (fd, error);
+}
+
+static gboolean
+fu_plugin_thunderbolt_power_reset_cb (gpointer data)
+{
+	if (!fu_plugin_thunderbolt_power_set (FU_PLUGIN (data), FALSE, NULL))
+		g_warning ("failed to reset thunderbolt power");
+	return FALSE;
+}
+
+static gboolean
+udev_uevent_cb (GUdevClient *udev,
+		const gchar *action,
+		GUdevDevice *device,
+		gpointer     user_data)
+{
+	FuPlugin *plugin = FU_PLUGIN(user_data);
+
+	if (action == NULL)
+		return TRUE;
+
+	g_debug ("uevent for %s: %s", g_udev_device_get_sysfs_path (device), action);
+
+	/* intel-wmi-thunderbolt has been loaded/unloaded */
+	if (g_str_equal (action, "change")) {
+		fu_plugin_thunderbolt_power_get_path (plugin);
+		fu_plugin_set_enabled (plugin, TRUE);
+		fu_plugin_recoldplug (plugin);
+	}
+
+	return TRUE;
+}
+
+/* virtual functions */
+
+void
+fu_plugin_init (FuPlugin *plugin)
+{
+	FuPluginData *data = fu_plugin_alloc_data (plugin, sizeof (FuPluginData));
+	const gchar *subsystems[] = { "wmi", NULL };
+
+	data->udev = g_udev_client_new (subsystems);
+	g_signal_connect (data->udev, "uevent",
+			  G_CALLBACK (udev_uevent_cb), plugin);
+	/* initially set to true, will wait for a device_register to reset */
+	data->needs_forcepower = TRUE;
+
+	/* determines whether to run device_registered */
+	fu_plugin_thunderbolt_power_get_path (plugin);
+}
+
+void
+fu_plugin_destroy (FuPlugin *plugin)
+{
+	FuPluginData *data = fu_plugin_get_data (plugin);
+	g_object_unref (data->udev);
+	if (data->force_path)
+		g_free (data->force_path);
+}
+
+void
+fu_plugin_device_registered (FuPlugin *plugin, FuDevice *device)
+{
+	FuPluginData *data = fu_plugin_get_data (plugin);
+
+	/* thunderbolt plugin */
+	if (g_strcmp0 (fu_device_get_plugin (device), "thunderbolt") == 0 &&
+	    fu_plugin_thunderbolt_power_supported (plugin)) {
+		if (data->needs_forcepower)
+			data->needs_forcepower = FALSE;
+		if (fu_device_has_flag (device, FWUPD_DEVICE_FLAG_INTERNAL)) {
+			fu_device_set_metadata_boolean (device,
+							FU_DEVICE_METADATA_TBT_CAN_FORCE_POWER,
+							TRUE);
+		}
+	}
+}
+
+gboolean
+fu_plugin_update_prepare (FuPlugin *plugin,
+			  FuDevice *device,
+			  GError **error)
+{
+	FuPluginData *data = fu_plugin_get_data (plugin);
+	g_autoptr(GUdevDevice) udevice = NULL;
+	const gchar *devpath;
+
+	/* only run for thunderbolt plugin */
+	if (g_strcmp0 (fu_device_get_plugin (device), "thunderbolt") != 0)
+		return TRUE;
+
+	devpath = fu_device_get_metadata (device, "sysfs-path");
+
+	udevice = g_udev_client_query_by_sysfs_path (data->udev, devpath);
+	if (udevice != NULL) {
+		data->needs_forcepower = FALSE;
+		return TRUE;
+	}
+	if (!fu_plugin_thunderbolt_power_set (plugin, TRUE, error))
+		return FALSE;
+
+	data->needs_forcepower = TRUE;
+	g_usleep (TBT_NEW_DEVICE_TIMEOUT * G_USEC_PER_SEC);
```

Comment:
Can you add a comment here explaining why the sleep is required please.

Reply:
Fixed